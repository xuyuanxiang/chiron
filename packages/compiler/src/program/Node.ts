export interface Node {
  readonly type: string;

  readonly [prop: string]: any;
}
