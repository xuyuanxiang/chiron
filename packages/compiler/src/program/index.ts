export * from './App';
export * from './Component';
export * from './File';
export * from './Node';
export * from './Package';
export * from './Page';
export * from './Program';
