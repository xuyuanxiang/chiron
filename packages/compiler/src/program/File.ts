import { createHash } from 'crypto';
import { Node } from './Node';

export enum ContentType {
  JS = 'js',
  WXSS = 'wxss',
  WXML = 'wxml',
  WXS = 'wxs',
  CONFIG = 'json',
}

export type DependencyMap = { [relativePath: string]: File };

export class File implements Node {
  readonly type: string = 'File';
  readonly start: number = 0;
  readonly end: number;
  readonly md5: string;
  readonly filename: string;
  readonly content: string;
  readonly contentType: ContentType;
  readonly dependencies?: DependencyMap;

  constructor(filename: string, content: string, contentType: ContentType) {
    const hash = createHash('md5');
    hash.update(content);
    this.md5 = hash.digest('hex');
    this.filename = filename;
    this.content = content;
    this.contentType = contentType;
    this.end = content.length;
    this.dependencies = {};
  }
}
