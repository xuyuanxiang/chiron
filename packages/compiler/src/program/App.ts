import { Node } from './Node';
import { File } from './File';

export class App implements Node {
  readonly type: string = 'App';
  readonly config: File;
  readonly script: File;

  constructor(config: File, script: File) {
    this.config = config;
    this.script = script;
  }
}
