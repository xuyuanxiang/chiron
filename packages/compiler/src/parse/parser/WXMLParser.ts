import { parseFragment, DocumentFragment } from 'parse5';
import { Parser } from './Parser';
import { ContentType } from '../../program';

export class WXMLParser implements Parser<DocumentFragment> {
  async parse(content: string): Promise<DocumentFragment> {
    return parseFragment(content);
  }

  support(contentType: ContentType): boolean {
    return contentType === ContentType.WXML;
  }
}
