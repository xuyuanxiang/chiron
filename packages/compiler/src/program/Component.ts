import { Page } from './Page';
import { File } from './File';
import { Node } from './Node';

export type ReferenceMap = { [alias: string]: Page };
export type ComponentDependencyMap = { [tag: string]: Component };

export interface ComponentConstructorArguments {
  script: File;
  template: File;
  style?: File;
  config?: File;
  references?: ReferenceMap;
  dependencies?: ComponentDependencyMap;
}

export class Component implements Node {
  readonly type: string = 'Component';
  readonly script: File;
  readonly template: File;
  readonly style?: File;
  readonly config?: File;
  readonly references?: ReferenceMap;
  readonly dependencies?: ComponentDependencyMap;

  constructor({
    script,
    template,
    style,
    config,
    references,
    dependencies,
  }: ComponentConstructorArguments) {
    this.script = script;
    this.template = template;
    this.style = style;
    this.config = config;
    this.references = references;
    this.dependencies = dependencies;
  }
}
