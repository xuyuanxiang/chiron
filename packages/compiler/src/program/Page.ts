import { Component, ComponentConstructorArguments } from './Component';

export interface PageConstructorArguments
  extends ComponentConstructorArguments {
  route: string;
}

export class Page extends Component {
  readonly type: string = 'Page';
  readonly route: string;

  constructor({ route, ...others }: PageConstructorArguments) {
    super(others);
    this.route = route;
  }
}
