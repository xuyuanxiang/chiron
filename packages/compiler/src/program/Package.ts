import { Node } from './Node';
import { Page } from './Page';

export class Package implements Node {
  readonly type: string = 'Package';
  readonly root: string;
  readonly pages: Page[];

  constructor(root: string, pages: Page[]) {
    this.root = root;
    this.pages = pages;
  }
}
