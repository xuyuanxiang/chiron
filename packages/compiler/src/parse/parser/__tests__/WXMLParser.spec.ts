import { WXMLParser } from '../WXMLParser';
import { ContentType } from '../../../program';

let parser: WXMLParser;

beforeEach(() => {
  parser = new WXMLParser();
});

it('should parse WXML', async () => {
  expect(parser.support(ContentType.WXML)).toBeTruthy();
  expect(parser.support(ContentType.CONFIG)).toBeFalsy();
  expect(parser.support(ContentType.WXSS)).toBeFalsy();
  expect(parser.support(ContentType.WXS)).toBeFalsy();
  expect(parser.support(ContentType.JS)).toBeFalsy();

  let ast = await parser.parse(`<view></view>`);
  expect(ast).toMatchSnapshot();
});
