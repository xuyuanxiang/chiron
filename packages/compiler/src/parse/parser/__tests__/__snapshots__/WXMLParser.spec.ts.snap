// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse WXML 1`] = `
Object {
  "childNodes": Array [
    Object {
      "attrs": Array [],
      "childNodes": Array [],
      "namespaceURI": "http://www.w3.org/1999/xhtml",
      "nodeName": "view",
      "parentNode": [Circular],
      "tagName": "view",
    },
  ],
  "nodeName": "#document-fragment",
}
`;
